sandbox/cgroup/kill.go

```diff
@@ -0,0 +1,122 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+// KillSnapProcesses sends a signal to all the processes belonging to
+// a given snap.
+//
+// A correct implementation is picked depending on cgroup v1 or v2 use in the
+// system. When cgroup v1 is detected, the call will directly act on the freezer
+// group created when a snap process was started, while with v2 the call will
+// act on all tracking groups of a snap.
+//
+// XXX: It is important to note that the implementation for v2 is racy because
+// processes can be killed externally even when the cgroup is frozen and have
+// their pid reused. Also, v2 freezing support is a no-op on kernels before
+// 5.2 which means new processes can keep getting spawned while killing
+// pids read earlier.
+var KillSnapProcesses = killSnapProcessesImplV1
+
+func pickKillV1Impl() {
+	KillSnapProcesses = killSnapProcessesImplV1
+}
+
+func pickKillV2Impl() {
+	KillSnapProcesses = killSnapProcessesImplV2
+}
+
+var syscallKill = syscall.Kill
+
+// killProcessesInCgroup sends signal to all the processes belonging to
+// passed cgroup directory.
+//
+// The caller is responsible for making sure that pids are not reused
+// after reading `cgroup.procs` to avoid TOCTOU.
+func killProcessesInCgroup(dir string, signal syscall.Signal) error {
+	pids, err := pidsInFile(filepath.Join(dir, "cgroup.procs"))
+	if err != nil {
+		return err
+	}
+
+	var firstErr error
+	for _, pid := range pids {
+		err := syscallKill(pid, signal)
+		if err != nil && err != syscall.ESRCH && firstErr == nil {
+			firstErr = err
+		}
+	}
+
+	return firstErr
+}
+
+func killSnapProcessesImplV1(ctx context.Context, snapName string, signal syscall.Signal) error {
+	if err := freezeSnapProcessesImplV1(ctx, snapName); err != nil {
+		return err
+	}
+	defer thawSnapProcessesImplV1(snapName)
```

Comment:
I think thaw may be duplicated in one of the inner functions. If we thaw here, let's drop the thaw from the lower level functions (or vice-versa but given that we freeze here it looks better).

Reply:
Separately, should we only thaw on error?

Reply:
`freezeSnapProcessesImplV1` internally only thaws if there is an error to avoid having a side effect on failure. 

For V1 we must thaw a frozen cgroup for the SIGKILL to take effect. It might be a good idea to document this kernel behavior in a comment.

Reply:
Ah, that is a very good point. Please leave a comment that makes this clear in the code.